Retry middleware

teandr opened the pull request.

teandr committed:
feat: create RetryMiddleware

teandr committed:
fix: make RequestSchedulerInterface a singleton

RequestSchedulerInterface was not registered as singleton.
Engine and RetryMiddleware received different scheduler instances.
This resulted in separate queues instead of a shared one.

teandr committed:
test: create RetryMiddlewareTest

ksassnowski commented on src/Downloader/Middleware/RetryMiddleware.php:
```diff
@@ -0,0 +1,84 @@
+
+            $retryRequest = $request
+                ->withMeta('retry_count', $retryCount + 1)
+                ->addOption('delay', $delay);
```
As far as I can tell, this delay isn't actually being used anywhere so the request will be retried immediately in the next batch.

teandr replied:
The delay will be made guzzle, immediately before the request.
https://docs.guzzlephp.org/en/stable/request-options.html#delay

ksassnowski commented on src/Downloader/Middleware/RetryMiddleware.php:
```diff
@@ -0,0 +1,84 @@
+            'retryOnStatus' => [500, 502, 503, 504],
+            'maxRetries' => 3,
+            'initialDelay' => 1000,
+            'delayMultiplier' => 2.0,
```
I would prefer if we combined `initialDelay` and `delayMultiplier` into a single `backoff` option that takes an array of integers.

So a backoff of `[1, 5, 10]` would mean the first retry happens after one second, the second after 5, the third after 10 seconds and 10 seconds for every subsequent request if `maxRetries` is greater than 3.

teandr replied:
Yes, good idea. Implemented.

ksassnowski commented:
Thanks for the PR! Looks good overall but I have a few comments that need to be addressed first. Please check them out when you find the time.

teandr committed:
refactor: refactor RetryMiddleware to use backoff option

teandr committed:
refactor: refactor RetryMiddlewareTest to use backoff option

teandr commented:
Hi, Kai.

Hope you're having a good week.

I'm writing to follow up on the pull request. I've addressed your previous comments and implemented the array-based backoff strategy, which is a much cleaner approach – thank you for the suggestion.

The implementation now allows a backoff option (e.g., [1, 5, 10]) to define the retry delays in seconds.

It then pushes this value into the request's options under the delay key.

Guzzle itself, as per its documentation, is responsible for honoring this option and pausing before sending the next request.

This approach delegates the actual waiting to the HTTP client, which is a best practice.

Would you mind reviewing the updated implementation when you get a chance? I believe it now cleanly combines the flexible configuration you suggested with the robust execution provided by Guzzle.

Please let me know if you have any further questions or if there's anything else you'd like to see adjusted.

Thank you for your time !
